File: java/twelve-days/src/main/java/TwelveDays.java
import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class TwelveDays {
    private static final String VERSE_FORMAT = "On the %s day of Christmas my true love gave to me: %s.\n";
    private static final Map<Integer, Verse> daysOfChristmas = new HashMap<Integer, Verse>() {{
        put(1, new Verse("first", "a Partridge in a Pear Tree"));
        put(2, new Verse("second", "two Turtle Doves"));
        put(3, new Verse("third", "three French Hens"));
        put(4, new Verse("fourth", "four Calling Birds"));
        put(5, new Verse("fifth", "five Gold Rings"));
        put(6, new Verse("sixth", "six Geese-a-Laying"));
        put(7, new Verse("seventh", "seven Swans-a-Swimming"));
        put(8, new Verse("eighth", "eight Maids-a-Milking"));
        put(9, new Verse("ninth", "nine Ladies Dancing"));
        put(10, new Verse("tenth", "ten Lords-a-Leaping"));
        put(11, new Verse("eleventh", "eleven Pipers Piping"));
        put(12, new Verse("twelfth", "twelve Drummers Drumming"));
    }};

    String verse(int verseNumber)
    {
        if (verseNumber == 1)
            return String.format(VERSE_FORMAT, daysOfChristmas.get(1).getNumber(), daysOfChristmas.get(1).getVerse());

        String verse = IntStream.rangeClosed(1, verseNumber).boxed().sorted(Comparator.reverseOrder())
                .map(this::buildVerse)
                .collect(Collectors.joining(", "));

        return String.format(VERSE_FORMAT, daysOfChristmas.get(verseNumber).getNumber(), verse);
    }

    String verses(int startVerse, int endVerse)
    {
        return IntStream.rangeClosed(startVerse, endVerse)
                .mapToObj(this::verse)
                .collect(Collectors.joining("\n"));
    }

    String sing()
    {
        return verses(1, 12);
    }

    private String buildVerse(int verseNumber)
    {
        return (verseNumber == 1 ? "and " : "") + daysOfChristmas.get(verseNumber).getVerse();
    }
}

class Verse
{
    private String number;
    private String verse;

    public Verse(String number, String verse)
    {
        this.number = number;
        this.verse = verse;
    }

    public String getNumber()
    {
        return number;
    }

    public String getVerse()
    {
        return verse;
    }
}

File: python/twelve-days/twelve_days.py
class Verse:
	def __init__(self, number: str, verse: str):
		self.number = number
		self.verse = verse

days_of_christmas = {
	1: Verse('first', 'a Partridge in a Pear Tree'),
	2: Verse('second', 'two Turtle Doves'),
	3: Verse('third', 'three French Hens'),
	4: Verse('fourth', 'four Calling Birds'),
	5: Verse('fifth', 'five Gold Rings'),
	6: Verse('sixth', 'six Geese-a-Laying'),
	7: Verse('seventh', 'seven Swans-a-Swimming'),
	8: Verse('eighth', 'eight Maids-a-Milking'),
	9: Verse('ninth', 'nine Ladies Dancing'),
	10: Verse('tenth', 'ten Lords-a-Leaping'),
	11: Verse('eleventh', 'eleven Pipers Piping'),
	12: Verse('twelfth', 'twelve Drummers Drumming')
}

verse_format = 'On the {} day of Christmas my true love gave to me: {}.'


def recite(start_verse: int, end_verse: int) -> list:
	return [build_verse(verse_number) for verse_number in range(start_verse, end_verse + 1)]


def build_verse(verse_number: int) -> str:
	if verse_number == 1:
		return verse_format.format(days_of_christmas[verse_number].number, days_of_christmas[verse_number].verse)

	line = ', '.join([days_of_christmas[num].verse if num != 1 else 'and {}'.format(days_of_christmas[num].verse) for num in range(verse_number, 0, -1)])

	return verse_format.format(days_of_christmas[verse_number].number, line)
